pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+
+    @classmethod
+    def _find_line_matches(cls, line: bytes, codec: str) -> Dict[int, _BadChar]:
+        """Find all matches of BAD_CHARS within line
+
+        Args:
+            line: the input
+            codec: that will be used to convert line/or search string into
+
+        Return:
+            An dictionary with the column offset and the BadASCIIChar
+        """
+        # We try to decode in Unicode to get the correct column offset
+        # if we would use bytes, it could be off because UTF-8 has no fixed length
+        try:
+            line_search = line.decode(codec, errors="strict")
+            search_dict = cls._BAD_ASCII_SEARCH_DICT
+            # If everything is correctly decoded, the BOM will be removed automatically
+            return _map_positions_to_result(line_search, search_dict, "\n")
+        except UnicodeDecodeError:
+            # If we can't decode properly, we simply use bytes, even so the column offsets
+            # might be wrong a bit, but it is still better then nothing
+            line_search_byte = line
+            search_dict_byte = {
+                _cached_encode_search(char.unescaped, codec): char
+                for char in cls._BAD_CHARS
+            }
+            # Using bytes we have to remove the BOM manually
+            return _map_positions_to_result(
+                line_search_byte,
+                search_dict_byte,
+                _cached_encode_search("\n", codec),
+                byte_str_length=_byte_to_str_length(codec),
+            )
+
+    def process_module(self, node: nodes.Module) -> None:
+        """Perform the actual check by checking module stream
+        the module's content is accessible via node.stream() function
+        """
+
+        # Check for UTF-16/UTF-32, to warn for it
+        lineno: int
+        line: bytes
+        codec: Optional[str]
+        lines: Sequence[bytes]
+        invalid_codec_line: Optional[int] = None
+
+        # First try to detect encoding with PEP 263, won't work
+        # see https://bugs.python.org/issue1503789
+        # pylint: disable-next=fixme
+        # TODO we could manually modify the stream to convert UTF16/UTF32 to ASCII
+        #      but why ;-) ?
+        try:
+            with node.stream() as stream:
+                codec, lines = detect_encoding(stream.readline)
+        except SyntaxError:
+            pass
+        else:
+            if self._is_invalid_codec(codec):
+                invalid_codec_line = len(lines)
+
+        # If there is no invalid codec already, lets try to determine a BOM,
+        # even so Python currently doesn't like non UTF-8 BOMs ....
+        if invalid_codec_line is None:
+            with node.stream() as stream:
+                line = stream.readline()
+                codec = extract_codec_from_bom(line)
+            if self._is_invalid_codec(codec):
+                invalid_codec_line = 1
+
+        if invalid_codec_line is not None:
```

Comment:
Is `len(lines)` ever 0? If not this could be simplified to `if invalid_codec_line`

Reply:
Same goes for the `is None`.

Reply:
Restructured the code, not required anymore